app/models/subscription/BusinessSubscriptionDetailsModel.scala

```diff
@@ -0,0 +1,198 @@
+}
+
+case class BusinessSubscriptionDetailsModel(accountingPeriod: AccountingPeriodModel,
+                                            selfEmploymentsData: Seq[SelfEmploymentData],
+                                            accountingMethod: AccountingMethod,
+                                            incomeSource: FeIncomeSourceModel,
+                                            propertyCommencementDate: Option[PropertyCommencementDateModel] = None,
+                                            propertyAccountingMethod: Option[AccountingMethodPropertyModel] = None,
+                                            overseasPropertyCommencementDate: Option[OverseasPropertyCommencementDateModel] = None,
+                                            overseasAccountingMethodProperty: Option[OverseasAccountingMethodPropertyModel] = None
+                                           )
+
+object BusinessSubscriptionDetailsModel {
+
+  implicit val businessSubscriptionReads: Reads[BusinessSubscriptionDetailsModel] = Json.reads[BusinessSubscriptionDetailsModel]
+
+  implicit val businessSubscriptionWrites: Writes[BusinessSubscriptionDetailsModel] = new Writes[BusinessSubscriptionDetailsModel] {
+    def writes(details: BusinessSubscriptionDetailsModel) = Json.obj(
+      "businessDetails" -> details.selfEmploymentsData.map(
+        data => Map("accountingPeriodStartDate" -> JsString(details.accountingPeriod.startDate.toDesDateFormat),
+          "accountingPeriodEndDate" -> JsString(details.accountingPeriod.endDate.toDesDateFormat),
+          "tradingName" -> JsString(data.businessTradeName.map(_.businessTradeName).get),
+          "addressDetails" -> Json.obj(
+            "addressLine1" -> data.businessAddress.map(_.address.lines.head).get,
+            "addressLine2" -> data.businessAddress.map(model => if(model.address.lines.length > 1) model.address.lines(1) else ""),
+            "addressLine3" -> data.businessAddress.map(model => if(model.address.lines.length > 2) model.address.lines(2) else ""),
+            "addressLine4" -> data.businessAddress.map(model => if(model.address.lines.length > 3) model.address.lines(3) else ""),
+            "postalCode" -> data.businessAddress.map(_.address.postcode).get,
+            "countryCode" -> "GB"
+          ),
+          "typeOfBusiness" -> JsString("self-employment"),
+          "tradingStartDate" -> JsString(data.businessStartDate.get.startDate.toDesDateFormat),
+          "cashOrAccrualsFlag" -> JsString(details.accountingMethod.stringValue)
+        )),
+      "ukPropertyDetails" -> Json.toJson(
+        if(details.incomeSource.ukProperty)
+          Json.obj(
+          "tradingStartDate" -> details.propertyCommencementDate.map(_.startDate.toDesDateFormat),
+          "cashOrAccrualsFlag" -> details.propertyAccountingMethod.map(_.propertyAccountingMethod.stringValue),
+          "startDate" -> details.accountingPeriod.startDate.toDesDateFormat
+        )
+        else JsNull
```

Comment:
This will not work. JsNull is not the same as a field being optional and will result in a failure. The field must not appear at all.

Reply:
Filtered Null and empty values.